src/jdk.jlink/share/classes/jdk/tools/jlink/internal/Snippets.java

```diff
@@ -0,0 +1,323 @@
+
+        protected CollectionSnippetBuilder(ClassDesc elementType) {
+            this.elementType = Objects.requireNonNull(elementType);
+        }
+
+        /**
+         * @param activatePagingThreshold  Use pagination methods if the count of elements is larger
+         *                                 than the given value
+         */
+        public CollectionSnippetBuilder activatePagingThreshold(int activatePagingThreshold) {
+            this.activatePagingThreshold = activatePagingThreshold;
+            return this;
+        }
+
+        /**
+         * @param ownerClassDesc  The owner class for the paginattion methods
+         */
+        public CollectionSnippetBuilder ownerClassDesc(ClassDesc ownerClassDesc) {
+            this.ownerClassDesc = ownerClassDesc;
+            return this;
+        }
+
+        /**
+         * @param methodNamePrefix  The method name prefix. Generated method will have the name of
+         *                          this value appended with page number
+         */
+        public CollectionSnippetBuilder methodNamePrefix(String methodNamePrefix) {
+            this.methodNamePrefix = methodNamePrefix;
+            return this;
+        }
+
+        /**
+         * @param pageSize  The count of elements per page*
+         */
+        public CollectionSnippetBuilder pageSize(int pageSize) {
+            this.pageSize = pageSize;
+            return this;
+        }
+
+        public CollectionSnippetBuilder classBuilder(ClassBuilder clb) {
+            this.clb = clb;
+            return this;
+        }
+
+        protected boolean shouldPaginate(int length) {
+            return activatePagingThreshold != 0 && length > activatePagingThreshold;
+        }
+
+        abstract public Loadable build(Snippet[] loadElementSnippets);
+    }
+
+    /**
+     * Generate bytecode for pagination methods, then create the array inline and invoke the first page method to assign
+     * values to the array. Each pagination method will assign value to the corresponding page and chain calling next
+     * page.
+     * {@code setup} must be called to generate the pagination methods in the owner class. Otherwise, {@code load} will
+     * lead to {@link java.lang.NoSuchMethodException}
+     *
+     * Effectively as
+     *   methodNamePrefix0(new T[elements.size()]);
+     *
+     * where
+     *   T[] methodNamePrefix0(T[] ar) {
+     *      ar[0] = elements[0];
+     *      ar[1] = elements[1];
+     *      ...
+     *      ar[pageSize-1] = elements[pageSize - 1];
+     *      methodNamePrefix1(ar);
+     *      return ar;
+     *   }
+     * and the last page will stop the chain and can be partial instead of full page size.
+     */
+    public static class ArraySnippetBuilder extends CollectionSnippetBuilder {
+        final MethodTypeDesc MTD_PageHelper;
+        final ClassDesc classDesc;
+        Snippet[] loadElementSnippets;
+
+        public ArraySnippetBuilder(ClassDesc elementType) {
```

Comment:
If pagination is activated, the caller should set up the pagination threshold, page size, method name prefix.  What happens if page size or method name prefix is not set?

One alternative is to have 2 constructors - the default constructor for no pagination and another constructor takes several required parameters for pagination for example 
```
        // return a builder with pagination activated and default threshold be PAGING_THRESHOLD
        public ArraySnippetBuilder(ClassDesc elementType, String methodNamePrefix, int pageSize) {
            ...
        }
```

 Can call `pagingThreshold(newThreshold)` to set a different threshold.

Reply:
I am planning to provides default value for those and validation when build.
Same applies to classBuilder, as it's optional and depending on whether helper methods is needed.